// @devvit/protos/types/devvit/ui/effects/v1alpha/app_permission.js
/**
 * #app_permission.ts
 *
 * Code generated by ts-proto. DO NOT EDIT.
 * @packageDocumentation
 */
/* eslint-disable */
import _m0 from "protobufjs/minimal.js";
import { consentStatusFromJSON, consentStatusToJSON, } from "../../../../reddit/devvit/app_permission/v1/app_permission.js";
import { messageTypeRegistry } from "../../../../typeRegistry.js";
function createBaseCanRunAsUserEffect() {
    return {};
}
export const CanRunAsUserEffect = {
    $type: "devvit.ui.effects.v1alpha.CanRunAsUserEffect",
    encode(_, writer = _m0.Writer.create()) {
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : _m0.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseCanRunAsUserEffect();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(_) {
        return {};
    },
    toJSON(_) {
        const obj = {};
        return obj;
    },
    create(base) {
        return CanRunAsUserEffect.fromPartial(base ?? {});
    },
    fromPartial(_) {
        const message = createBaseCanRunAsUserEffect();
        return message;
    },
};
messageTypeRegistry.set(CanRunAsUserEffect.$type, CanRunAsUserEffect);
function createBaseConsentStatusEvent() {
    return { consentStatus: 0 };
}
export const ConsentStatusEvent = {
    $type: "devvit.ui.effects.v1alpha.ConsentStatusEvent",
    encode(message, writer = _m0.Writer.create()) {
        if (message.consentStatus !== 0) {
            writer.uint32(8).int32(message.consentStatus);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : _m0.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseConsentStatusEvent();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 8) {
                        break;
                    }
                    message.consentStatus = reader.int32();
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return { consentStatus: isSet(object.consentStatus) ? consentStatusFromJSON(object.consentStatus) : 0 };
    },
    toJSON(message) {
        const obj = {};
        if (message.consentStatus !== 0) {
            obj.consentStatus = consentStatusToJSON(message.consentStatus);
        }
        return obj;
    },
    create(base) {
        return ConsentStatusEvent.fromPartial(base ?? {});
    },
    fromPartial(object) {
        const message = createBaseConsentStatusEvent();
        message.consentStatus = object.consentStatus ?? 0;
        return message;
    },
};
messageTypeRegistry.set(ConsentStatusEvent.$type, ConsentStatusEvent);
function isSet(value) {
    return value !== null && value !== undefined;
}
