// @devvit/protos/json/devvit/dev_portal/installation/installation.js
/**
 * #installation.ts
 *
 * Code generated by ts-proto. DO NOT EDIT.
 * @packageDocumentation
 */
/* eslint-disable */
export var UpgradeStrategy;
(function (UpgradeStrategy) {
    UpgradeStrategy[UpgradeStrategy["MANUAL"] = 0] = "MANUAL";
    UpgradeStrategy[UpgradeStrategy["AUTOMATIC"] = 1] = "AUTOMATIC";
    UpgradeStrategy[UpgradeStrategy["UNRECOGNIZED"] = -1] = "UNRECOGNIZED";
})(UpgradeStrategy || (UpgradeStrategy = {}));
export var InstallationHistoryEventName;
(function (InstallationHistoryEventName) {
    InstallationHistoryEventName[InstallationHistoryEventName["INSTALLED"] = 0] = "INSTALLED";
    InstallationHistoryEventName[InstallationHistoryEventName["UNINSTALLED"] = 1] = "UNINSTALLED";
    InstallationHistoryEventName[InstallationHistoryEventName["UPGRADED"] = 2] = "UPGRADED";
    InstallationHistoryEventName[InstallationHistoryEventName["UPDATED"] = 3] = "UPDATED";
    InstallationHistoryEventName[InstallationHistoryEventName["BANNED"] = 4] = "BANNED";
    InstallationHistoryEventName[InstallationHistoryEventName["UNBANNED"] = 5] = "UNBANNED";
    InstallationHistoryEventName[InstallationHistoryEventName["UNRECOGNIZED"] = -1] = "UNRECOGNIZED";
})(InstallationHistoryEventName || (InstallationHistoryEventName = {}));
