// @devvit/client/effects/share.js
import { emitEffect } from '@devvit/shared-types/client/emit-effect.js';
import { maxShareParamUserDataChars } from '@devvit/shared-types/client/share.js';
import { ICON_FILE_PATH } from '@devvit/shared-types/constants.js';
// to-do: unit test.
// to-do: move to web view scripts.
export async function showShareSheet(opts) {
    if (opts.data && opts.data.length > maxShareParamUserDataChars)
        throw Error(`data must be <= ${maxShareParamUserDataChars} characters but was ${opts.data.length} characters`);
    const iconURL = `${new URL(ICON_FILE_PATH, location.origin)}`;
    let iconRsp;
    try {
        iconRsp = await fetch(iconURL, { method: 'HEAD' });
    }
    catch {
        //
    }
    await emitEffect({
        type: 9,
        share: {
            appIconUri: iconRsp?.ok ? iconURL : undefined,
            userData: opts.data,
            text: opts.text,
            title: opts.title,
        },
    });
}
export function getShareData() {
    return devvit.share?.userData;
}
