// @devvit/protos/types/devvit/plugin/v1alpha/pushnotif/pushnotif.twirp.js
import { TwirpServer, TwirpError, TwirpErrorCode, TwirpContentType, chainInterceptors, } from 'twirp-ts';
import { QueuePushNotificationRequest, QueuePushNotificationResponse, } from './pushnotif.js';
export var PushNotifMethod;
(function (PushNotifMethod) {
    PushNotifMethod["QueuePushNotification"] = "QueuePushNotification";
})(PushNotifMethod || (PushNotifMethod = {}));
export const PushNotifMethodList = [PushNotifMethod.QueuePushNotification];
export function createPushNotifServer(service) {
    return new TwirpServer({
        service,
        packageName: 'devvit.plugin.v1alpha.pushnotif',
        serviceName: 'PushNotif',
        methodList: PushNotifMethodList,
        matchRoute: matchPushNotifRoute,
    });
}
function matchPushNotifRoute(method, events) {
    switch (method) {
        case 'QueuePushNotification':
            return async (ctx, service, data, interceptors) => {
                ctx = { ...ctx, methodName: 'QueuePushNotification' };
                await events.onMatch(ctx);
                return handlePushNotifQueuePushNotificationRequest(ctx, service, data, interceptors);
            };
        default:
            events.onNotFound();
            const msg = `no handler found`;
            throw new TwirpError(TwirpErrorCode.BadRoute, msg);
    }
}
function handlePushNotifQueuePushNotificationRequest(ctx, service, data, interceptors) {
    switch (ctx.contentType) {
        case TwirpContentType.JSON:
            return handlePushNotifQueuePushNotificationJSON(ctx, service, data, interceptors);
        case TwirpContentType.Protobuf:
            return handlePushNotifQueuePushNotificationProtobuf(ctx, service, data, interceptors);
        default:
            const msg = 'unexpected Content-Type';
            throw new TwirpError(TwirpErrorCode.BadRoute, msg);
    }
}
async function handlePushNotifQueuePushNotificationJSON(ctx, service, data, interceptors) {
    let request;
    let response;
    try {
        const body = JSON.parse(data.toString() || '{}');
        request = QueuePushNotificationRequest.fromJSON(body);
    }
    catch (e) {
        if (e instanceof Error) {
            const msg = 'the json request could not be decoded';
            throw new TwirpError(TwirpErrorCode.Malformed, msg).withCause(e, true);
        }
    }
    if (interceptors && interceptors.length > 0) {
        const interceptor = chainInterceptors(...interceptors);
        response = await interceptor(ctx, request, (ctx, inputReq) => {
            return service.QueuePushNotification(ctx, inputReq);
        });
    }
    else {
        response = await service.QueuePushNotification(ctx, request);
    }
    return JSON.stringify(QueuePushNotificationResponse.toJSON(response));
}
async function handlePushNotifQueuePushNotificationProtobuf(ctx, service, data, interceptors) {
    let request;
    let response;
    try {
        request = QueuePushNotificationRequest.decode(data);
    }
    catch (e) {
        if (e instanceof Error) {
            const msg = 'the protobuf request could not be decoded';
            throw new TwirpError(TwirpErrorCode.Malformed, msg).withCause(e, true);
        }
    }
    if (interceptors && interceptors.length > 0) {
        const interceptor = chainInterceptors(...interceptors);
        response = await interceptor(ctx, request, (ctx, inputReq) => {
            return service.QueuePushNotification(ctx, inputReq);
        });
    }
    else {
        response = await service.QueuePushNotification(ctx, request);
    }
    return Buffer.from(QueuePushNotificationResponse.encode(response).finish());
}
