// @devvit/cli/dist/util/Bundler.js
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _Bundler_buildPack;
import { ESBuildPack } from '@devvit/build-pack/esbuild/ESBuildPack.js';
import { formatLogs } from '@devvit/build-pack/lib/BuildPack.js';
import {} from '@devvit/protos/types/devvit/plugin/buildpack/buildpack_common.js';
import { LOCAL_HOSTNAME } from '@devvit/shared-types/HostnameUtil.js';
import { map } from 'rxjs';
import { DEVVIT_DISABLE_EXTERN_DEVVIT_PROTOS } from './config.js';
export class Bundler {
    constructor() {
        _Bundler_buildPack.set(this, void 0);
        __classPrivateFieldSet(this, _Bundler_buildPack, new ESBuildPack({ hostname: LOCAL_HOSTNAME }, { disableExternDevvitProtos: DEVVIT_DISABLE_EXTERN_DEVVIT_PROTOS }), "f");
    }
    async bundle(project, actorSpec, includeMetafile = false) {
        const compiledRes = await __classPrivateFieldGet(this, _Bundler_buildPack, "f").compile(
        // to-do: why no minify?
        {
            config: project.appConfig,
            info: actorSpec,
            minify: 'None',
            includeMetafile,
            root: project.root,
        });
        if (compiledRes.warnings.length > 0) {
            console.warn(formatLogs(compiledRes.warnings));
        }
        if (compiledRes.errors.length > 0) {
            throw new Error(formatLogs(compiledRes.errors));
        }
        if (compiledRes.bundles.length === 0) {
            throw new Error('Missing bundle');
        }
        return compiledRes.bundles;
    }
    async dispose() {
        await __classPrivateFieldGet(this, _Bundler_buildPack, "f").dispose();
    }
    watch(project, root, actorSpec) {
        return __classPrivateFieldGet(this, _Bundler_buildPack, "f")
            .watch({
            config: project?.appConfig,
            info: actorSpec,
            minify: 'None',
            includeMetafile: false,
            root,
        })
            .pipe(map((rsp) => {
            if (rsp.warnings.length > 0)
                console.warn(formatLogs(rsp.warnings));
            if (rsp.errors.length > 0)
                console.error(formatLogs(rsp.errors));
            return { bundles: rsp.bundles };
        }));
    }
}
_Bundler_buildPack = new WeakMap();
