// @devvit/reddit/plugin.js
import { FlairDefinition, GraphQLDefinition, LinksAndCommentsDefinition, ListingsDefinition, ModerationDefinition, ModNoteDefinition, NewModmailDefinition, PrivateMessagesDefinition, SubredditsDefinition, UsersDefinition, WidgetsDefinition, WikiDefinition, } from '@devvit/protos';
import { UserActionsDefinition } from '@devvit/protos';
import { getDevvitConfig } from '@devvit/shared-types/server/get-devvit-config.js';
let redditApiPlugin;
let userActionsPlugin;
/**
 * @internal
 */
export function getRedditApiPlugins() {
    return (redditApiPlugin ?? (redditApiPlugin = {
        Flair: getDevvitConfig().use(FlairDefinition),
        GraphQL: getDevvitConfig().use(GraphQLDefinition),
        LinksAndComments: getDevvitConfig().use(LinksAndCommentsDefinition),
        Listings: getDevvitConfig().use(ListingsDefinition),
        ModNote: getDevvitConfig().use(ModNoteDefinition),
        Moderation: getDevvitConfig().use(ModerationDefinition),
        NewModmail: getDevvitConfig().use(NewModmailDefinition),
        PrivateMessages: getDevvitConfig().use(PrivateMessagesDefinition),
        Subreddits: getDevvitConfig().use(SubredditsDefinition),
        Users: getDevvitConfig().use(UsersDefinition),
        Widgets: getDevvitConfig().use(WidgetsDefinition),
        Wiki: getDevvitConfig().use(WikiDefinition),
    }));
}
/**
 * @internal
 */
export function getUserActionsPlugin() {
    if (!getDevvitConfig().uses(UserActionsDefinition)) {
        throw new Error(`UserActions plugin is not enabled. To use 'runAs: "USER"', set 'permissions.reddit.asUser: [ "PERMISSION_SCOPE_NAME" ]' in your devvit.json file.`);
    }
    return (userActionsPlugin ?? (userActionsPlugin = getDevvitConfig().use(UserActionsDefinition)));
}
