// @devvit/cli/dist/commands/uninstall.js
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _Uninstall_installationsClient, _Uninstall_appVersionClient;
import { InstallationType } from '@devvit/protos/community.js';
import { StringUtil } from '@devvit/shared-types/StringUtil.js';
import { Args, ux } from '@oclif/core';
import { getAccessTokenAndLoginIfNeeded } from '../util/auth.js';
import { createAppVersionClient, createInstallationsClient } from '../util/clientGenerators.js';
import { DevvitCommand, toLowerCaseArgParser } from '../util/commands/DevvitCommand.js';
import { getSubredditNameWithoutPrefix } from '../util/common-actions/getSubredditNameWithoutPrefix.js';
class Uninstall extends DevvitCommand {
    constructor() {
        super(...arguments);
        _Uninstall_installationsClient.set(this, createInstallationsClient());
        _Uninstall_appVersionClient.set(this, createAppVersionClient());
    }
    async run() {
        const { args } = await this.parse(Uninstall);
        const subreddit = getSubredditNameWithoutPrefix(args.subreddit);
        const appName = args.app || this.project.name;
        await getAccessTokenAndLoginIfNeeded('LocalSocket');
        ux.action.start(`Finding installation of the app "${appName}" in r/${subreddit}`);
        let id;
        try {
            const allInstalledHere = await __classPrivateFieldGet(this, _Uninstall_installationsClient, "f").GetAllWithInstallLocation({
                location: subreddit,
                type: InstallationType.SUBREDDIT,
            });
            // TODO I do not like how I'm doing this. I think we should make a new API call for this, but this _does_ work for now.
            const installationToRemove = (await Promise.all(allInstalledHere.installations.map(async (installation) => {
                const fullInstallInfo = await __classPrivateFieldGet(this, _Uninstall_installationsClient, "f").GetByUUID({
                    id: installation.id,
                });
                const appVersionInfo = await __classPrivateFieldGet(this, _Uninstall_appVersionClient, "f").Get({
                    id: fullInstallInfo.appVersion?.id ?? '',
                });
                if (appVersionInfo.app?.slug === appName) {
                    // This is the one we want to uninstall
                    return installation;
                }
                else {
                    return null;
                }
            }))).find((maybeInstall) => !!maybeInstall);
            if (!installationToRemove) {
                this.error(`Could not uninstall - no installation of "${appName}" exists in that location!`);
            }
            id = installationToRemove.id;
        }
        catch (err) {
            ux.action.stop('Error');
            this.error(`Error while trying to find installation to remove: ${StringUtil.caughtToString(err, 'message')}`);
        }
        ux.action.stop();
        ux.action.start(`Uninstalling`);
        try {
            await __classPrivateFieldGet(this, _Uninstall_installationsClient, "f").Remove({ id });
        }
        catch (err) {
            ux.action.stop('Error');
            this.error(`ERROR: Unable to uninstall app: ${StringUtil.caughtToString(err, 'message')}`);
        }
        ux.action.stop();
    }
}
_Uninstall_installationsClient = new WeakMap(), _Uninstall_appVersionClient = new WeakMap();
Uninstall.description = 'Uninstall an app from the specified subreddit';
Uninstall.examples = [
    '$ devvit uninstall <subreddit> [app]',
    '$ devvit uninstall r/myTestSubreddit',
    '$ devvit uninstall myOtherTestSubreddit my-app',
];
Uninstall.args = {
    subreddit: Args.string({
        description: 'Provide the name of the subreddit to uninstall from. The "r/" prefix is optional',
        required: true,
        parse: toLowerCaseArgParser,
    }),
    app: Args.string({
        description: 'Provide the name of the app to uninstall (defaults to working directory app).',
        required: false,
        parse: toLowerCaseArgParser,
    }),
};
export default Uninstall;
