// @devvit/reddit/models/WikiPage.js
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _a, _WikiPage_name, _WikiPage_subredditName, _WikiPage_content, _WikiPage_contentHtml, _WikiPage_revisionId, _WikiPage_revisionDate, _WikiPage_revisionReason, _WikiPage_revisionAuthor, _WikiPage_metadata_get, _WikiPageRevision_id, _WikiPageRevision_page, _WikiPageRevision_date, _WikiPageRevision_author, _WikiPageRevision_reason, _WikiPageRevision_hidden, _WikiPageSettings_listed, _WikiPageSettings_permLevel, _WikiPageSettings_editors;
import { context } from '@devvit/server';
import { assertNonNull } from '@devvit/shared-types/NonNull.js';
import { makeGettersEnumerable } from '../helpers/makeGettersEnumerable.js';
import { getRedditApiPlugins } from '../plugin.js';
import { Listing } from './Listing.js';
import { User } from './User.js';
export var WikiPagePermissionLevel;
(function (WikiPagePermissionLevel) {
    /** Use subreddit wiki permissions */
    WikiPagePermissionLevel[WikiPagePermissionLevel["SUBREDDIT_PERMISSIONS"] = 0] = "SUBREDDIT_PERMISSIONS";
    /** Only approved wiki contributors for this page may edit */
    WikiPagePermissionLevel[WikiPagePermissionLevel["APPROVED_CONTRIBUTORS_ONLY"] = 1] = "APPROVED_CONTRIBUTORS_ONLY";
    /** Only mods may edit and view */
    WikiPagePermissionLevel[WikiPagePermissionLevel["MODS_ONLY"] = 2] = "MODS_ONLY";
})(WikiPagePermissionLevel || (WikiPagePermissionLevel = {}));
export class WikiPage {
    /**
     * @internal
     */
    constructor(name, subredditName, data) {
        _WikiPage_name.set(this, void 0);
        _WikiPage_subredditName.set(this, void 0);
        _WikiPage_content.set(this, void 0);
        _WikiPage_contentHtml.set(this, void 0);
        _WikiPage_revisionId.set(this, void 0);
        _WikiPage_revisionDate.set(this, void 0);
        _WikiPage_revisionReason.set(this, void 0);
        _WikiPage_revisionAuthor.set(this, void 0);
        makeGettersEnumerable(this);
        __classPrivateFieldSet(this, _WikiPage_name, name, "f");
        __classPrivateFieldSet(this, _WikiPage_subredditName, subredditName, "f");
        __classPrivateFieldSet(this, _WikiPage_content, data.contentMd, "f");
        __classPrivateFieldSet(this, _WikiPage_contentHtml, data.contentHtml, "f");
        __classPrivateFieldSet(this, _WikiPage_revisionId, data.revisionId, "f");
        __classPrivateFieldSet(this, _WikiPage_revisionDate, new Date(data.revisionDate * 1000), "f"); // data.revisionDate is represented in seconds, so multiply by 1000 to get milliseconds
        __classPrivateFieldSet(this, _WikiPage_revisionReason, data.reason ?? '', "f");
        __classPrivateFieldSet(this, _WikiPage_revisionAuthor, data.revisionBy?.data ? new User(data.revisionBy.data) : undefined, "f");
    }
    /** The name of the page. */
    get name() {
        return __classPrivateFieldGet(this, _WikiPage_name, "f");
    }
    /** The name of the subreddit the page is in. */
    get subredditName() {
        return __classPrivateFieldGet(this, _WikiPage_subredditName, "f");
    }
    /** The Markdown content of the page. */
    get content() {
        return __classPrivateFieldGet(this, _WikiPage_content, "f");
    }
    /** The HTML content of the page. */
    get contentHtml() {
        return __classPrivateFieldGet(this, _WikiPage_contentHtml, "f");
    }
    /** The ID of the revision. */
    get revisionId() {
        return __classPrivateFieldGet(this, _WikiPage_revisionId, "f");
    }
    /** The date of the revision. */
    get revisionDate() {
        return __classPrivateFieldGet(this, _WikiPage_revisionDate, "f");
    }
    /** The reason for the revision. */
    get revisionReason() {
        return __classPrivateFieldGet(this, _WikiPage_revisionReason, "f");
    }
    /** The author of this revision. */
    get revisionAuthor() {
        return __classPrivateFieldGet(this, _WikiPage_revisionAuthor, "f");
    }
    toJSON() {
        return {
            name: __classPrivateFieldGet(this, _WikiPage_name, "f"),
            subredditName: __classPrivateFieldGet(this, _WikiPage_subredditName, "f"),
            content: __classPrivateFieldGet(this, _WikiPage_content, "f"),
            contentHtml: __classPrivateFieldGet(this, _WikiPage_contentHtml, "f"),
            revisionId: __classPrivateFieldGet(this, _WikiPage_revisionId, "f"),
            revisionDate: __classPrivateFieldGet(this, _WikiPage_revisionDate, "f"),
            revisionReason: __classPrivateFieldGet(this, _WikiPage_revisionReason, "f"),
            revisionAuthor: __classPrivateFieldGet(this, _WikiPage_revisionAuthor, "f")?.toJSON(),
        };
    }
    /** Update this page. */
    async update(content, reason) {
        return _a.updatePage({
            subredditName: __classPrivateFieldGet(this, _WikiPage_subredditName, "f"),
            page: __classPrivateFieldGet(this, _WikiPage_name, "f"),
            content,
            reason,
        });
    }
    /** Get the revisions for this page. */
    async getRevisions(options) {
        return _a.getPageRevisions({
            subredditName: __classPrivateFieldGet(this, _WikiPage_subredditName, "f"),
            page: __classPrivateFieldGet(this, _WikiPage_name, "f"),
            ...options,
        });
    }
    /** Revert this page to a previous revision. */
    async revertTo(revisionId) {
        return _a.revertPage(__classPrivateFieldGet(this, _WikiPage_subredditName, "f"), __classPrivateFieldGet(this, _WikiPage_name, "f"), revisionId);
    }
    /** Get the settings for this page. */
    async getSettings() {
        return _a.getPageSettings(__classPrivateFieldGet(this, _WikiPage_subredditName, "f"), __classPrivateFieldGet(this, _WikiPage_name, "f"));
    }
    /** Update the settings for this page. */
    async updateSettings(options) {
        return _a.updatePageSettings({
            subredditName: __classPrivateFieldGet(this, _WikiPage_subredditName, "f"),
            page: __classPrivateFieldGet(this, _WikiPage_name, "f"),
            listed: options.listed,
            permLevel: options.permLevel,
        });
    }
    /** Add an editor to this page. */
    async addEditor(username) {
        return _a.addEditor(__classPrivateFieldGet(this, _WikiPage_subredditName, "f"), __classPrivateFieldGet(this, _WikiPage_name, "f"), username);
    }
    /** Remove an editor from this page. */
    async removeEditor(username) {
        return _a.removeEditor(__classPrivateFieldGet(this, _WikiPage_subredditName, "f"), __classPrivateFieldGet(this, _WikiPage_name, "f"), username);
    }
    /** @internal */
    static async getPage(subredditName, page) {
        const client = getRedditApiPlugins().Wiki;
        const response = await client.GetWikiPage({
            subreddit: subredditName,
            page,
        }, __classPrivateFieldGet(this, _a, "a", _WikiPage_metadata_get));
        assertNonNull(response.data, 'Failed to get wiki page');
        return new _a(page, subredditName, response.data);
    }
    /** @internal */
    static async getPages(subredditName) {
        const client = getRedditApiPlugins().Wiki;
        const response = await client.GetWikiPages({ subreddit: subredditName }, __classPrivateFieldGet(this, _a, "a", _WikiPage_metadata_get));
        return response.data || [];
    }
    /** @internal */
    static async createPage(options) {
        return _a.updatePage(options);
    }
    /** @internal */
    static async updatePage(options) {
        const client = getRedditApiPlugins().Wiki;
        await client.EditWikiPage({
            subreddit: options.subredditName,
            page: options.page,
            content: options.content,
            reason: options.reason ?? '',
        }, __classPrivateFieldGet(this, _a, "a", _WikiPage_metadata_get));
        return _a.getPage(options.subredditName, options.page);
    }
    /** @internal */
    static getPageRevisions(options) {
        const client = getRedditApiPlugins().Wiki;
        return new Listing({
            hasMore: true,
            after: options.after,
            limit: options.limit,
            pageSize: options.pageSize,
            fetch: async (fetchOptions) => {
                const response = await client.GetWikiPageRevisions({
                    subreddit: options.subredditName,
                    page: options.page,
                    limit: fetchOptions.limit,
                    after: fetchOptions.after,
                    before: fetchOptions.before,
                }, __classPrivateFieldGet(this, _a, "a", _WikiPage_metadata_get));
                return wikiPageRevisionListingProtoToWikiPageRevision(response);
            },
        });
    }
    /** @internal */
    static async revertPage(subredditName, page, revisionId) {
        const client = getRedditApiPlugins().Wiki;
        await client.RevertWikiPage({
            subreddit: subredditName,
            page,
            revision: revisionId,
        }, __classPrivateFieldGet(this, _a, "a", _WikiPage_metadata_get));
    }
    /** @internal */
    static async getPageSettings(subredditName, page) {
        const client = getRedditApiPlugins().Wiki;
        const response = await client.GetWikiPageSettings({
            subreddit: subredditName,
            page,
        }, __classPrivateFieldGet(this, _a, "a", _WikiPage_metadata_get));
        assertNonNull(response.data, 'Failed to get wiki page settings');
        return new WikiPageSettings(response.data);
    }
    /** @internal */
    static async updatePageSettings(options) {
        const client = getRedditApiPlugins().Wiki;
        const response = await client.UpdateWikiPageSettings({
            subreddit: options.subredditName,
            page: options.page,
            listed: options.listed ? 'on' : '',
            permlevel: options.permLevel,
        }, __classPrivateFieldGet(this, _a, "a", _WikiPage_metadata_get));
        assertNonNull(response.data, 'Failed to update wiki page settings');
        return new WikiPageSettings(response.data);
    }
    /** @internal */
    static async addEditor(subredditName, page, username) {
        const client = getRedditApiPlugins().Wiki;
        await client.AllowEditor({
            act: 'add',
            subreddit: subredditName,
            page,
            username,
        }, __classPrivateFieldGet(this, _a, "a", _WikiPage_metadata_get));
    }
    /** @internal */
    static async removeEditor(subredditName, page, username) {
        const client = getRedditApiPlugins().Wiki;
        await client.AllowEditor({
            act: 'del',
            subreddit: subredditName,
            page,
            username,
        }, __classPrivateFieldGet(this, _a, "a", _WikiPage_metadata_get));
    }
}
_a = WikiPage, _WikiPage_name = new WeakMap(), _WikiPage_subredditName = new WeakMap(), _WikiPage_content = new WeakMap(), _WikiPage_contentHtml = new WeakMap(), _WikiPage_revisionId = new WeakMap(), _WikiPage_revisionDate = new WeakMap(), _WikiPage_revisionReason = new WeakMap(), _WikiPage_revisionAuthor = new WeakMap(), _WikiPage_metadata_get = function _WikiPage_metadata_get() {
    return context.metadata;
};
export class WikiPageRevision {
    constructor(data) {
        _WikiPageRevision_id.set(this, void 0);
        _WikiPageRevision_page.set(this, void 0);
        _WikiPageRevision_date.set(this, void 0);
        _WikiPageRevision_author.set(this, void 0);
        _WikiPageRevision_reason.set(this, void 0);
        _WikiPageRevision_hidden.set(this, void 0);
        __classPrivateFieldSet(this, _WikiPageRevision_id, data.id, "f");
        __classPrivateFieldSet(this, _WikiPageRevision_page, data.page, "f");
        __classPrivateFieldSet(this, _WikiPageRevision_date, new Date(data.timestamp), "f");
        assertNonNull(data.author?.data, 'Wiki page revision author details are missing');
        __classPrivateFieldSet(this, _WikiPageRevision_author, new User(data.author.data), "f");
        __classPrivateFieldSet(this, _WikiPageRevision_reason, data.reason ?? '', "f");
        __classPrivateFieldSet(this, _WikiPageRevision_hidden, data.revisionHidden ?? false, "f");
    }
    get id() {
        return __classPrivateFieldGet(this, _WikiPageRevision_id, "f");
    }
    get page() {
        return __classPrivateFieldGet(this, _WikiPageRevision_page, "f");
    }
    get date() {
        return __classPrivateFieldGet(this, _WikiPageRevision_date, "f");
    }
    get author() {
        return __classPrivateFieldGet(this, _WikiPageRevision_author, "f");
    }
    get reason() {
        return __classPrivateFieldGet(this, _WikiPageRevision_reason, "f");
    }
    get hidden() {
        return __classPrivateFieldGet(this, _WikiPageRevision_hidden, "f");
    }
    toJSON() {
        return {
            id: __classPrivateFieldGet(this, _WikiPageRevision_id, "f"),
            page: __classPrivateFieldGet(this, _WikiPageRevision_page, "f"),
            date: __classPrivateFieldGet(this, _WikiPageRevision_date, "f"),
            author: __classPrivateFieldGet(this, _WikiPageRevision_author, "f").toJSON(),
            reason: __classPrivateFieldGet(this, _WikiPageRevision_reason, "f"),
            hidden: __classPrivateFieldGet(this, _WikiPageRevision_hidden, "f"),
        };
    }
}
_WikiPageRevision_id = new WeakMap(), _WikiPageRevision_page = new WeakMap(), _WikiPageRevision_date = new WeakMap(), _WikiPageRevision_author = new WeakMap(), _WikiPageRevision_reason = new WeakMap(), _WikiPageRevision_hidden = new WeakMap();
export class WikiPageSettings {
    constructor(data) {
        _WikiPageSettings_listed.set(this, void 0);
        _WikiPageSettings_permLevel.set(this, void 0);
        _WikiPageSettings_editors.set(this, void 0);
        __classPrivateFieldSet(this, _WikiPageSettings_listed, data.listed, "f");
        __classPrivateFieldSet(this, _WikiPageSettings_permLevel, data.permLevel, "f");
        __classPrivateFieldSet(this, _WikiPageSettings_editors, data.editors.map((editor) => {
            assertNonNull(editor.data, 'Wiki page editor details are missing');
            return new User(editor.data);
        }), "f");
    }
    get listed() {
        return __classPrivateFieldGet(this, _WikiPageSettings_listed, "f");
    }
    get permLevel() {
        return __classPrivateFieldGet(this, _WikiPageSettings_permLevel, "f");
    }
    get editors() {
        return __classPrivateFieldGet(this, _WikiPageSettings_editors, "f");
    }
    toJSON() {
        return {
            listed: __classPrivateFieldGet(this, _WikiPageSettings_listed, "f"),
            permLevel: __classPrivateFieldGet(this, _WikiPageSettings_permLevel, "f"),
            editors: __classPrivateFieldGet(this, _WikiPageSettings_editors, "f").map((editor) => editor.toJSON()),
        };
    }
}
_WikiPageSettings_listed = new WeakMap(), _WikiPageSettings_permLevel = new WeakMap(), _WikiPageSettings_editors = new WeakMap();
function wikiPageRevisionListingProtoToWikiPageRevision(listingProto) {
    assertNonNull(listingProto.data?.children, 'Wiki page revision listing is missing children');
    const children = listingProto.data.children.map((child) => {
        return new WikiPageRevision(child);
    });
    return {
        children,
        before: listingProto.data.before,
        after: listingProto.data.after,
    };
}
